server/bleep/src/indexes/doc.rs

```diff
@@ -0,0 +1,997 @@
+    EmptyDocs(url::Url),
+}
+
+impl Doc {
+    fn index_writer(&self) -> Result<tantivy::IndexWriter, Error> {
+        self.section_index
+            .writer(crate::config::default_buffer_size())
+            .map_err(Error::Tantivy)
+    }
+
+    fn index_reader(&self) -> Result<tantivy::IndexReader, Error> {
+        self.section_index.reader().map_err(Error::Tantivy)
+    }
+    /// Initialize docs index
+    pub async fn create(sql: SqlDb, path: &std::path::Path) -> Result<Self, Error> {
+        std::fs::create_dir_all(path)?;
+
+        let section_schema = SectionSchema::new();
+        let mut section_index = tantivy::Index::open_or_create(
+            tantivy::directory::MmapDirectory::open(path)
+                .map_err(|e| Error::Initialize(e.to_string()))?, // todo: handle tantivy index upgrades here
+            section_schema.schema.clone(),
+        )
+        .map_err(|e| Error::Initialize(e.to_string()))?;
+
+        section_index
+            .set_default_multithread_executor()
+            .map_err(|e| Error::Initialize(e.to_string()))?;
+        section_index
+            .set_multithread_executor(crate::config::default_parallelism())
+            .map_err(|e| Error::Initialize(e.to_string()))?;
+
+        section_index
+            .tokenizers()
+            .register("trigram", NgramTokenizer::new(1, 3, false).unwrap());
+
+        // Ok(index)
+
+        Ok(Self {
+            sql,
+            section_index,
+            section_schema,
+        })
+    }
+
+    async fn set_title(&self, title: &str, id: i64) -> Result<(), Error> {
+        sqlx::query! {
+            "UPDATE docs SET name = ? WHERE id = ?",
+            title,
+            id,
+        }
+        .execute(&*self.sql)
+        .await
+        .map(|_| ())
+        .map_err(Error::Sql)
+    }
+
+    async fn set_favicon(&self, favicon: &str, id: i64) -> Result<(), Error> {
+        sqlx::query! {
+            "UPDATE docs SET favicon = ? WHERE id = ?",
+            favicon,
+            id,
+        }
+        .execute(&*self.sql)
+        .await
+        .map(|_| ())
+        .map_err(Error::Sql)
+    }
+
+    async fn set_description(&self, description: &str, id: i64) -> Result<(), Error> {
+        sqlx::query! {
+            "UPDATE docs SET description = ? WHERE id = ?",
+            description,
+            id,
+        }
+        .execute(&*self.sql)
+        .await
+        .map(|_| ())
+        .map_err(Error::Sql)
+    }
+
+    async fn set_index_status(&self, status: &str, id: i64) -> Result<(), Error> {
+        let status = status.to_string();
+        sqlx::query! {
+            "UPDATE docs SET index_status = ? WHERE id = ?",
+            status,
+            id,
+        }
+        .execute(&*self.sql)
+        .await
+        .map(|_| ())
+        .map_err(Error::Sql)
+    }
+
+    /// Add a doc source to the index
+    pub async fn sync(self, url: url::Url) -> impl Stream<Item = Result<Progress, Error>> {
```

Comment:
Can you add a comment how tantivy and sqlite are used in this process at a high level, just in a single sentence?